pandas/core/reshape/concat.py

```diff
@@ -221,23 +223,35 @@ def concat(objs, axis=0, join='outer', join_axes=None, ignore_index=False,
         ...
     ValueError: Indexes have overlapping values: ['a']
     """
-    op = _Concatenator(objs, axis=axis, join_axes=join_axes,
-                       ignore_index=ignore_index, join=join,
+    op = _Concatenator(objs, axis=axis, ignore_index=ignore_index, join=join,
                        keys=keys, levels=levels, names=names,
-                       verify_integrity=verify_integrity,
-                       copy=copy, sort=sort)
-    return op.get_result()
+                       verify_integrity=verify_integrity, copy=copy, sort=sort)
+
+    res = op.get_result()
+
+    if join_axes is not None:
+        # GH 21951
+        warnings.warn('The join_axes-keyword is deprecated. Use .reindex or '
+                      '.reindex_like on the result to achieve the same '
+                      'functionality.', FutureWarning, stacklevel=2)
+        ndim = res.ndim
+        if len(join_axes) != ndim - 1:
+            raise AssertionError("join_axes must be a list of indexes of "
+                                 "length {length}".format(length=ndim - 1))
+        if ndim == 2:
+            other_axis = 1 if axis == 0 else 0  # switches between 0 & 1
+            res = res.reindex(join_axes[0], axis=other_axis, copy=False)
```

Comment:
Should we move this reindexing to before passing the objs to `_Concatenator` ?

Reply:
@h-vetinari can you update for this point